presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1331,10 +1333,16 @@ public LayoutConstraintEvaluator(Session session, Map<Symbol, Type> types, Map<S
             Map<NodeRef<Expression>, Type> expressionTypes = getExpressionTypes(session, metadata, parser, types, expression, emptyList());
 
             evaluator = ExpressionInterpreter.expressionOptimizer(expression, metadata, session, expressionTypes);
+            arguments = SymbolsExtractor.extractUnique(expression).stream()
+                    .map(assignments::get)
+                    .collect(toImmutableSet());
         }
 
         private boolean isCandidate(Map<ColumnHandle, NullableValue> bindings)
         {
+            if (!arguments.stream().anyMatch(bindings.keySet()::contains)) {
```

Comment:
Isn't that `bindings.keySet()` will be the same in every call to `isCandidate()`?

Reply:
for cases I am aware of it is true, but it is not enforced anywhere and connector may pass anything it want

Reply:
Shouldn't it be `allMatch` instead?